---
title: "Pangenome statistics"
author: "Liam Shaw"
date: "`r Sys.Date()`"
output: 
  html_document:
  fig_width: 12
fig_height: 8
editor_options: 
  chunk_output_type: console
---
  
  
  
```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
dataDir = "../data/" # assumes being run from 'notebooks' dir
figureDir = "../output-figures/" # for figures
outputDir = "../output-data/" # for output data

# Pangenome component colours
component.colours = c("#1f78b4", "#a6cee3", "#b2df8a")
names(component.colours) = c("Core", "Non-core", "Plasmid")
```

## Libraries

```{r libraries, include=FALSE}
library(dplyr)
library(ggplot2)
library(cowplot)
library(formatR)
library(ape)
library(ggtree)
library(ggridges)
library(MCMCglmm)
library(phytools)
library(reshape2)
library(tidyr)
library(ggrepel)
```

```{r}
pangenome.stats = read.csv(paste0(dataDir, "pangenome-stats.csv"),
                header=T,
                stringsAsFactors=F)
pangenome.stats$mean.genes.per.genome = pangenome.stats$total.genes/pangenome.stats$genomes
pangenome.stats$mean.plasmid.genes.per.genome = pangenome.stats$total.plasmid.genes/pangenome.stats$genomes
pangenome.stats$mean.nonplasmid.noncoregenes.per.genome = pangenome.stats$mean.genes.per.genome-pangenome.stats$mean.plasmid.genes.per.genome-pangenome.stats$core.genes

pangenome.stats.melt = melt(pangenome.stats, id.vars = "species", measure.vars = c("core.genes",
                                                                                   "mean.nonplasmid.noncoregenes.per.genome",
                                                                                   "mean.plasmid.genes.per.genome"))
pangenome.stats.melt$variable = ordered(pangenome.stats.melt$variable,
                                        levels=rev(c("core.genes",
                                                                                   "mean.nonplasmid.noncoregenes.per.genome",
                                                                                   "mean.plasmid.genes.per.genome")),
                                        labels=rev(c("Core", "Non-core", "Plasmid")))
```

```{r plot-pangenome}
ggplot(pangenome.stats.melt, aes(species, value, fill=variable))+
  geom_bar(stat="identity", position="stack")+
  theme_bw()+
  coord_flip()+
  scale_fill_manual(values=component.colours)+
  theme(panel.grid = element_blank())
```


```{r genomes}
contigs.db = read.csv(paste0(dataDir, "2022-11-18-contig-database.csv"), 
                      header=T,
                      stringsAsFactors = F,
                      row.names = 1)
View(contigs.db %>% filter(contig.number>1, plasmid!=TRUE))
  
```
